compaction/compaction.cc

```diff
@@ -1703,6 +1704,107 @@ future<uint64_t> scrub_validate_mode_validate_reader(flat_mutation_reader_v2 rea
+    uint64_t errors{0};
+    auto idx_reader = sstables::index_reader(sst, permit, pc, {}, sstables::use_caching::no);
+    const auto& schema = *sst->get_schema();
+    partition_key::tri_compare cmp(schema);
+
+    auto validate_partition = [&] (const partition_key& pkey) -> future<> {
+        std::exception_ptr ex;
+        auto reader = sst->make_reader(sst->get_schema(), permit, dht::partition_range::make_singular(dht::decorate_key(schema, pkey)), schema.full_slice(), pc,
+                {}, streamed_mutation::forwarding::no, mutation_reader::forwarding::no);
+        try {
+            auto maybe_ps_opt = co_await reader();
+            if (!maybe_ps_opt) {
+                ++errors;
+                clogger.error("[scrub compaction {}.{}] Invalid index entry for partition {} ({}), partition not found in data file",
+                        schema.ks_name(), schema.cf_name(), scrub_compaction::partition_key_to_string(pkey, schema), pkey);
+            } else if (!maybe_ps_opt->is_partition_start()) {
+                ++errors;
+                clogger.error("[scrub compaction {}.{}] Invalid index entry for partition {} ({}), entry doesn't point to start of partition",
+                        schema.ks_name(), schema.cf_name(), scrub_compaction::partition_key_to_string(pkey, schema), pkey);
+            } else {
+                mutation_fragment_v2_opt last_mf;
+                while (auto next_mf = co_await reader()) {
+                    last_mf = std::move(next_mf);
+                }
+                if (!last_mf || !last_mf->is_end_of_partition()) {
+                    ++errors;
+                    clogger.error("[scrub compaction {}.{}] Invalid index entry for partition {} ({}), partition incomplete",
+                            schema.ks_name(), schema.cf_name(), scrub_compaction::partition_key_to_string(pkey, schema), pkey);
+                }
+            }
+        } catch (...) {
+            ex = std::current_exception();
+        }
+        if (ex) {
+            std::rethrow_exception(std::move(ex));
+        }
+        co_await reader.close();
+    };
+
+    std::exception_ptr ex;
+    try {
+        co_await idx_reader.read_partition_data();
+        auto last_pkey = idx_reader.get_partition_key();
+        auto last_pos = idx_reader.get_data_file_position();
+
+        co_await validate_partition(last_pkey);
+
+        while (!idx_reader.eof()) {
+            if (cdata.is_stop_requested()) [[unlikely]] {
+                // Compaction manager will catch this exception and re-schedule the compaction.
+                throw compaction_stopped_exception(schema.ks_name(), schema.cf_name(), cdata.stop_requested);
+            }
+
+            co_await idx_reader.read_partition_data();
+            auto pos = idx_reader.get_data_file_position();
+            auto pkey = idx_reader.get_partition_key();
+
```

Comment:
This duplication can be avoided by using optional for last_pkey and last_pos, making the procedure easier to read.

Reply:
Yes, this duplication is a remnant of a time when I constructed the reader once and re-used it.